x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/registry/ModelRegistry.java

```diff
@@ -393,4 +433,57 @@ private static IndexRequest createIndexRequest(String docId, String indexName, T
     private QueryBuilder documentIdQuery(String inferenceEntityId) {
         return QueryBuilders.constantScoreQuery(QueryBuilders.idsQuery().addIds(Model.documentId(inferenceEntityId)));
     }
+
+    static UnparsedModel deepCopyDefaultConfig(UnparsedModel other) {
```

Comment:
Instead of this deep copying, wouldn't it a better to make `UnparsedModel` immutable, and have a Builder.

Reply:
Because the settings are service dependent only the service knows how to parse them. At this stage it stored as a loose map where the stored json has been parsed into a map of maps. This was a bad design decision but not one to be fixed here. `UnparsedModel` is a record so it is immutable but the maps it holds may not be. Another issue is that when the services read the configs they remove elements from the map so that they can check if there are any settings left in the map they do not know about, the maps must be _mutable_ for the individual fields to be removed. Another bad design decision. The deep copy is necessary because we have maps of maps. This is fragile code but it is well covered in tests. I hope to address the problems described here later.

Reply:
thanks for the explanation

Reply:
FWIW, I agree with "This was a bad design decision but not one to be fixed her".